pkg/distributor/distributor.go

```diff
@@ -2789,6 +2798,48 @@ respsLoop:
 	return result, nil
 }
 
+// adjustQueryRequestLimit recalculated the query request limit.
+// The returned value is the approximation, a query to an individual shard needs to be limited with.
+func (d *Distributor) adjustQueryRequestLimit(ctx context.Context, userID string, limit int) int {
+	if limit == 0 {
+		return limit
+	}
+
+	var shardSize int
+	if d.cfg.IngestStorageConfig.Enabled {
+		// Get the number of active partitions in the ring. Here the ShuffleShardSize handles cases when a tenant has 0 or negative
+		// number of shards, or more shards than the number of active partitions in the ring.
+		shardSize = d.partitionsRing.PartitionRing().ShuffleShardSize(d.limits.IngestionPartitionsTenantShardSize(userID))
+	} else {
+		tenantShardSize := d.limits.IngestionTenantShardSize(userID)
+
+		// The ShuffleShard filters out read-only instances, leaving us with the number of active ingesters.
+		subring := d.ingestersRing.ShuffleShard(userID, tenantShardSize)
+		ingestionShardSize := subring.InstancesWithTokensCount()
+		if tenantShardSize > 0 {
+			ingestionShardSize = min(tenantShardSize, ingestionShardSize)
```

Comment:
We already pass the `tenantShardSize` to `ShuffleShard()` above. Why do we need this logic here as well? Isn't `ingestionShardSize` already the actual shard size?

Reply:
Good call. Looking at it now, I think I was overly protective here for now reason. I've simplified the chunk.